src/System.Net.Security/src/System/Net/SecureProtocols/InternalNegoState.cs

```diff
@@ -0,0 +1,822 @@
+                if (policy.CustomChannelBinding == null && policy.CustomServiceNames == null)
+                {
+                    throw new ArgumentException(SR.net_auth_must_specify_extended_protection_scheme, "policy");
+                }
+
+                _extendedProtectionPolicy = policy;
+            }
+            else
+            {
+                _extendedProtectionPolicy = new ExtendedProtectionPolicy(PolicyEnforcement.Never);
+            }
+
+            ValidateCreateContext(package, true, credential, servicePrincipalName, _extendedProtectionPolicy.CustomChannelBinding, protectionLevel, impersonationLevel);
+        }
+
+        internal void ValidateCreateContext(
+            string package,
+            bool isServer,
+            NetworkCredential credential,
+            string servicePrincipalName,
+            ChannelBinding channelBinding,
+            ProtectionLevel protectionLevel,
+            TokenImpersonationLevel impersonationLevel)
+        {
+            if (_exception != null && !_canRetryAuthentication)
+            {
+                throw _exception;
+            }
+
+            if (_context != null && _context.IsValidContext)
+            {
+                throw new InvalidOperationException(SR.net_auth_reauth);
+            }
+
+            if (credential == null)
+            {
+                throw new ArgumentNullException("credential");
+            }
+
+            if (servicePrincipalName == null)
+            {
+                throw new ArgumentNullException("servicePrincipalName");
+            }
+
+            ValidateImpersonationLevel(impersonationLevel);
+
+            if (_context != null && IsServer != isServer)
+            {
+                throw new InvalidOperationException(SR.net_auth_client_server);
+            }
+
+            _exception = null;
+            _remoteOk = false;
+            _framer = new StreamFramer(_innerStream);
+            _framer.WriteHeader.MessageId = FrameHeader.HandshakeId;
+
+            _expectedProtectionLevel = protectionLevel;
+            _expectedImpersonationLevel = isServer ? impersonationLevel : TokenImpersonationLevel.None;
+            _writeSequenceNumber = 0;
+            _readSequenceNumber = 0;
+
+            ContextFlagsPal flags = ContextFlagsPal.Connection;
+
+            // A workaround for the client when talking to Win9x on the server side.
+            if (protectionLevel == ProtectionLevel.None && !isServer)
+            {
+                package = NegotiationInfoClass.NTLM;
+            }
+
+            else if (protectionLevel == ProtectionLevel.EncryptAndSign)
+            {
+                flags |= ContextFlagsPal.Confidentiality;
+            }
+            else if (protectionLevel == ProtectionLevel.Sign)
+            {
+                // Assuming user expects NT4 SP4 and above.
+                flags |= (ContextFlagsPal.ReplayDetect | ContextFlagsPal.SequenceDetect | ContextFlagsPal.InitIntegrity);
+            }
```

Comment:
`if`/`else if`/`else if`/... waterfalls that don't end in an `else` always make me nervous.  Right now `ProtectionLevel.None` with `isServer` set doesn't do anything.  Are there other cases that don't do anything?  If not, might it be reasonable to add

```C#
else
{
    Debug.Assert(protectionLevel == ProtectionLevel.None);
}
```

So if a future protection mode comes online (or somehow the value 93 accidentally gets passed in) we can catch it.